llvm/lib/Target/AIE/AIERegClassConstrainer.cpp

```diff
@@ -0,0 +1,135 @@
+#include "llvm/CodeGen/MachineRegisterInfo.h"
+#include "llvm/CodeGen/SlotIndexes.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+#include <map>
+
+#include "AIE.h"
+
+using namespace llvm;
+
+#define DEBUG_TYPE "aie-ra"
+
+namespace {
+
+/// Constrain partially allocatable classes to fully allocatable ones.
+/// This makes sure that greedy works with classes for which it is easy to
+/// generate spill code.
+class AIERegClassConstrainer : public MachineFunctionPass {
+
+public:
+  static char ID;
+  AIERegClassConstrainer() : MachineFunctionPass(ID) {}
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.setPreservesCFG();
+    AU.addPreserved<SlotIndexes>();
+    AU.addPreserved<LiveIntervals>();
+    MachineFunctionPass::getAnalysisUsage(AU);
+  }
+
+  bool runOnMachineFunction(MachineFunction &Fn) override;
+
+private:
+  /// Find the largest subclass of \p RC that is suitable for register
+  /// allocation. That is, it does not contain reserved registers, as those
+  /// are harder to spill/reload.
+  const TargetRegisterClass &getSuitableRCForRA(const TargetRegisterClass &RC,
+                                                const MachineRegisterInfo &MRI);
+
+  /// Cache for \p getSuitableRCForRA
+  std::map<unsigned, const TargetRegisterClass *> SuitableRCForRA;
+};
+
+bool AIERegClassConstrainer::runOnMachineFunction(MachineFunction &MF) {
+  LLVM_DEBUG(llvm::dbgs() << "Constraining RCs for RA: " << MF.getName()
+                          << "\n");
+  MachineRegisterInfo &MRI = MF.getRegInfo();
+  const TargetRegisterInfo &TRI = *MRI.getTargetRegisterInfo();
+
+  // Clear previous cache, reserved registers can change for each function
+  SuitableRCForRA.clear();
+
+  for (unsigned I = 0, E = MRI.getNumVirtRegs(); I != E; ++I) {
+    Register Reg = Register::index2VirtReg(I);
+    if (MRI.reg_nodbg_empty(Reg))
+      continue;
+
+    const TargetRegisterClass &SrcRC = *MRI.getRegClass(Reg);
+    const TargetRegisterClass &SuitableRC = getSuitableRCForRA(SrcRC, MRI);
+    if (SrcRC.getID() == SuitableRC.getID())
+      continue;
+
+    LLVM_DEBUG(llvm::dbgs() << "  Constraining " << printReg(Reg, &TRI)
+                            << " from " << TRI.getRegClassName(&SrcRC) << " to "
+                            << TRI.getRegClassName(&SuitableRC) << "\n");
+    MRI.setRegClass(Reg, &SuitableRC);
+  }
+
+  // We are only changing register classes, so it's probably okay to return
+  // false. But let's be conservative, the return value of runOnMachineFunction
+  // isn't very well documented.
+  return true;
+}
+
+const TargetRegisterClass &
+getSuitableRCForRAImpl(const TargetRegisterClass &RC,
+                       const MachineRegisterInfo &MRI) {
+  const TargetRegisterInfo &TRI = *MRI.getTargetRegisterInfo();
+
+  auto ContainsReservedReg = [&MRI](const TargetRegisterClass &RC) {
+    return any_of(RC, [&MRI](MCRegister Reg) { return MRI.isReserved(Reg); });
+  };
+
```

Comment:
If `const TargetRegisterClass &RC` does not contain reserved regs at all, do we need to iterate over all SubRegClass set?

Reply:
I think that if the class simply is free of reserved registers, then the loop will find the same class.

Reply:
Correct, I can add an early exit